hbase-server/src/test/java/org/apache/hadoop/hbase/quotas/TestSpaceQuotasWithRegionReplicas.java

```diff
@@ -0,0 +1,119 @@
+
+  @BeforeClass
+  public static void setUp() throws Exception {
+    Configuration conf = TEST_UTIL.getConfiguration();
+    SpaceQuotaHelperForTests.updateConfigForQuotas(conf);
+    TEST_UTIL.startMiniCluster(1);
+  }
+
+  @AfterClass
+  public static void tearDown() throws Exception {
+    TEST_UTIL.shutdownMiniCluster();
+  }
+
+  @Before
+  public void removeAllQuotas() throws Exception {
+    helper = new SpaceQuotaHelperForTests(TEST_UTIL, testName, new AtomicLong(0));
+    helper.removeAllQuotas();
+  }
+
+  @Test
+  public void testSetQuotaWithRegionReplicaSingleRegion() throws Exception {
+    setQuotaAndVerifyForRegionReplication(1, 2, SpaceViolationPolicy.NO_INSERTS);
+    setQuotaAndVerifyForRegionReplication(1, 2, SpaceViolationPolicy.NO_WRITES);
+    setQuotaAndVerifyForRegionReplication(1, 2, SpaceViolationPolicy.NO_WRITES_COMPACTIONS);
+    setQuotaAndVerifyForRegionReplication(1, 2, SpaceViolationPolicy.DISABLE);
+  }
+
+  @Test
+  public void testSetQuotaWithRegionReplicaMultipleRegion() throws Exception {
+    setQuotaAndVerifyForRegionReplication(5, 3, SpaceViolationPolicy.NO_INSERTS);
+    setQuotaAndVerifyForRegionReplication(6, 3, SpaceViolationPolicy.NO_WRITES);
+    setQuotaAndVerifyForRegionReplication(6, 3, SpaceViolationPolicy.NO_WRITES_COMPACTIONS);
+    setQuotaAndVerifyForRegionReplication(6, 3, SpaceViolationPolicy.DISABLE);
+  }
+
+  @Test
+  public void testSetQuotaWithSingleRegionZeroRegionReplica() throws Exception {
+    setQuotaAndVerifyForRegionReplication(1, 0, SpaceViolationPolicy.NO_INSERTS);
+    setQuotaAndVerifyForRegionReplication(1, 0, SpaceViolationPolicy.NO_WRITES);
+    setQuotaAndVerifyForRegionReplication(1, 0, SpaceViolationPolicy.NO_WRITES_COMPACTIONS);
+    setQuotaAndVerifyForRegionReplication(1, 0, SpaceViolationPolicy.DISABLE);
+  }
+
+  @Test
+  public void testSetQuotaWithMultipleRegionZeroRegionReplicas() throws Exception {
+    setQuotaAndVerifyForRegionReplication(5, 0, SpaceViolationPolicy.NO_INSERTS);
+    setQuotaAndVerifyForRegionReplication(6, 0, SpaceViolationPolicy.NO_WRITES);
+    setQuotaAndVerifyForRegionReplication(6, 0, SpaceViolationPolicy.NO_WRITES_COMPACTIONS);
+    setQuotaAndVerifyForRegionReplication(6, 0, SpaceViolationPolicy.DISABLE);
+  }
+
+  private void setQuotaAndVerifyForRegionReplication(int region, int replicatedRegion,
+      SpaceViolationPolicy policy) throws Exception {
+    TableName tn = helper.createTableWithRegions(TEST_UTIL.getAdmin(),
+        NamespaceDescriptor.DEFAULT_NAMESPACE_NAME_STR, region, replicatedRegion);
+    helper.setQuotaLimit(tn, policy, 5L);
+    helper.writeData(tn, 5L * SpaceQuotaHelperForTests.ONE_MEGABYTE);
+    Put p = new Put(Bytes.toBytes("to_reject"));
+    p.addColumn(Bytes.toBytes(SpaceQuotaHelperForTests.F1), Bytes.toBytes("to"),
+        Bytes.toBytes("reject"));
+    // Adding a sleep for 5 sec, so all the chores run and to void flakiness of the test.
+    Thread.sleep(5000);
```

Comment:
:-1: Make this an explicit check. What do you need to propagate so that this test can pass? The RegionServerQuotaManager needs to be aware of the quota that was set? You can look at that via `HBaseTestingUtility` and the `HRegionServer` object you can get from there.

Is that clear? Like your other PR, there should be examples of doing this already in the `o.a.h.h.quotas` package.

Reply:
Not required here since verifyViolation() has NUM_RETRIES as 10 and also has some wait. So removing the Thread.Sleep here. Even without it, testcase will always pass.